paper/paper.md

```diff
@@ -0,0 +1,48 @@
+---
+title: 'Setfield.jl: Changing the immutable'
+tags:
+  - Julia
+  - Functional Programming
+  - Optics
+authors:
+  - name: Jan Weidner
+    orcid: 0000-0002-0980-8239
+    affiliation: 1
+affiliations:
+ - name: PTW-Dosimetry
+   index: 1
+date: 13 September 2020
+bibliography: paper.bib
+---
+
+# Summary
+We discuss the problem of updating immutable objects. The solutions presented are implemented in the Setfield.jl package [@SetfieldPackage].
+
+# Overview
+
+In the Julia programming language [@Julia-2017], some objects are *mutable* (`Array`, `mutable struct`, `...`), while others are *immutable* (`Tuple`, `struct`, `...`).
+Neither is strictly better than the other in every situation. However, *immutability* usually leads to code that is easier to reason about, for both humans and compilers.
+And therefore less buggy and more performant programs.
+One convenience with mutability is, that it makes updating objects very simple:
+
+`spaceship.captain.name = "Julia"`
+
+The analogous operation in the immutable case is to create a copy of `spaceship`,
+with just the captains name changed to "Julia". This operation is sometimtes called functional update.
+Just think for a moment, how would you do achieve this?
+It is a non trivial problem and there are many approaches. Both in Julia [@JuliaPR21912; @MutabilitiesPackage] and other languages [@HaskellLens; @ImmutableJS].
```

Comment:
I've heard Clojure's Specter took a different(?) approach to lens for mutating immutables but I haven't looked into it yet carefully: http://nathanmarz.com/blog/clojures-missing-piece.html

Reply:
Cool, I will at take a look!

Reply:
Elm took an alternative approach: don't use nested immutable and make flat struct/record better.

https://elm-lang.org/docs/records (see "extensible records")
https://www.reddit.com/r/elm/comments/6c6f3b/elm_monocle_vs_haskell_lens_why_the_differences/

I think there was a talk by the creator of Elm in youtube briefly discussing this. But I can't find it at the moment.

Reply:
Maybe I can write a paragraph about this (so that I'd be more conformable to put me as an author :smile:) but I don't have enough time right now. Is there a deadline for the paper?

Reply:
AFAIU the deadline is somewhere mid october:



> Hello everyone,
We hope you had a fantastic JuliaCon 2020! We always enjoy learning about the breadth and drive of the community, as well as meeting old and new friends.
Last year we decided to have conference proceedings for the first time and we are happy and proud to announce that the first set of articles has been published! We are still working on the last few and they will be published on a rolling basis.
We would like to thank all the authors (for writing fantastic articles and for their patience), the reviewers (for their valuable contributions), and the Open Journals (especially Arfon Smith) for their infrastructure and technical support.
So what is next? Every speaker of this JuliaCon edition has the opportunity to write up an article corresponding to their talk and submit it to the JuliaCon proceedings. We accept extended abstracts (1 page without citations ) and full-sized papers (8 pages).
The important dates:
September 30th: article submission of at https://proceedings.juliacon.org Mid-october: Paper must be finalized & compilable. Authors should suggest a set of reviewers. October-June: Review period Published date: July 2021
We are still working on the website with all the details and information about the process.
If you are interested in volunteering as an Editor or Reviewer please contact one of us!
Kind regards,
Mathieu, Valentin & Ranjan JuliaCon proceedings editors

Reply:
Interestingly, it looks like Haskell/GHC recently decided to add a syntax for nested fields

https://github.com/ghc-proposals/ghc-proposals/pull/282
https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0282-record-dot-syntax.rst

(found it via https://news.ycombinator.com/item?id=24538531)

Reply:
Oh nice! Not being able to use the same names for fields of distinct types, was always very awkward in haskell.

Reply:
btw I took a look at Specter. I like it. I think it is close to traversals in "traditional" optics. Just uses different language. https://github.com/JuliaObjects/Accessors.jl/pull/6 is somewhat inspired by it.

Reply:
FYI I plan to submit the draft tonight (tomorrow in my time zone). I would still like very much to have you listed as a coauthor. If you don't have time to add to the paper text, that's not a concern I think. What counts is your contributions to the project and it feels somewhat strange to me to not have you listed.

Reply:
I just submitted the paper.